packages/message-relayer/test/unit-tests/relay-tx.spec.ts

```diff
@@ -0,0 +1,375 @@
+    it('should return null if a batch for the index does not exist', async () => {
+      // Should have a total of 1 element now.
+      await StateCommitmentChain.appendStateBatch(
+        [ethers.constants.HashZero],
+        0
+      )
+
+      expect(
+        await getStateRootBatchByTransactionIndex(
+          l1RpcProvider,
+          StateCommitmentChain.address,
+          1 // Index 0 is ok but 1 should return null
+        )
+      ).to.equal(null)
+    })
+
+    it('should return the full batch for a given index when it exists', async () => {
+      // Should have a total of 1 element now.
+      await StateCommitmentChain.appendStateBatch(
+        [ethers.constants.HashZero],
+        0
+      )
+
+      const batch = await getStateRootBatchByTransactionIndex(
+        l1RpcProvider,
+        StateCommitmentChain.address,
+        0 // Index 0 is ok but 1 should return null
+      )
+
+      expect(batch.header).to.deep.include({
+        batchIndex: ethers.BigNumber.from(0),
+        batchSize: ethers.BigNumber.from(1),
+        prevTotalElements: ethers.BigNumber.from(0),
+      })
+
+      expect(batch.stateRoots).to.deep.equal([ethers.constants.HashZero])
+    })
+  })
+
+  describe('makeRelayTransactionData', () => {
+    it('should throw an error if the transaction does not exist', async () => {
+      await expect(
+        makeRelayTransactionData(
+          l1RpcProvider,
+          l2RpcProvider,
+          StateCommitmentChain.address,
+          MockL2CrossDomainMessenger.address,
+          ethers.constants.HashZero
+        )
+      ).to.be.rejected
+    })
+
+    it('should throw an error if the transaction did not send a message', async () => {
+      const tx = await MockL2CrossDomainMessenger.doNothing()
+
+      await expect(
+        makeRelayTransactionData(
+          l1RpcProvider,
+          l2RpcProvider,
+          StateCommitmentChain.address,
+          MockL2CrossDomainMessenger.address,
+          tx.hash
+        )
+      ).to.be.rejected
+    })
+
+    it('should throw an error if the corresponding state batcsenderh has not been submitted', async () => {
+      const tx = await MockL2CrossDomainMessenger.emitSentMessageEvent({
+        target: ethers.constants.AddressZero,
+        sender: ethers.constants.AddressZero,
+        message: '0x',
+        messageNonce: 0,
+      })
+
+      await expect(
+        makeRelayTransactionData(
+          l1RpcProvider,
+          l2RpcProvider,
+          StateCommitmentChain.address,
+          MockL2CrossDomainMessenger.address,
+          tx.hash
+        )
+      ).to.be.rejected
+    })
+
+    // Unfortunately this is hard to test here because hardhat doesn't support eth_getProof.
+    // Because this function is embedded into the message relayer, we should be able to use
+    // integration tests to sufficiently test this.
+    it.skip('should otherwise return the encoded transaction data', () => {
+      // TODO?
```

Comment:
I've tested this manually (by relaying transactions on kovan), but we won't be able to fully integration test this until we've integrated this into the message-relayer. I started working on that but quickly realized that the diff would become *way* too large for just a single PR. So instead I propose that we review this with the knowledge that it definitely *does* work and that we'll get more confirmation of this fact in a soon-to-be-created PR.